test/celeritas/track/StatusChecker.test.cc

```diff
@@ -0,0 +1,233 @@
+        }
+        catch (RichContextException const& e)
+        {
+            CELER_LOG(info) << "Exception during action " << id.get()
+                            << " came from: " << e.what();
+            actual_message = "<Failed to rethrow rich context correctly>";
+            try
+            {
+                // Rethrow to print the underlying message
+                std::rethrow_if_nested(e);
+            }
+            catch (DebugError const& e)
+            {
+                actual_message = e.details().condition;
+            }
+        }
+        catch (DebugError const& e)
+        {
+            actual_message = e.details().condition;
+        }
+        catch (RuntimeError const& e)
+        {
+            // This gets thrown if a CUDA assert is triggered
+            actual_message = e.details().what;
+        }
+
+        EXPECT_TRUE(actual_message.find(match) != std::string::npos)
+            << "Actual message: '" << actual_message << "'";
+    }
+
+  private:
+    std::shared_ptr<StatusChecker> status_checker_;
+};
+
+TEST_F(StatusCheckerTest, host)
+{
+    CoreState<MemSpace::host> state{*this->core(), StreamId{0}, 16};
+    this->begin_run(state);
+    state.insert_primaries(make_span(this->make_primaries(8)));
+
+    // Keep a persistent view to the last track slot
+    SimTrackView sim(
+        core()->ref<MemSpace::host>().sim, state.ref().sim, TrackSlotId{15});
+
+    // Actions: see Stepper.test.cc
+    EXPECT_EQ(TrackStatus::inactive, sim.status());
+    auto id = this->find_action("extend-from-primaries");
+    this->run(id, state);
+    EXPECT_EQ(TrackStatus::inactive, sim.status());
+
+    id = this->find_action("initialize-tracks");
+    this->run(id, state);
+    EXPECT_EQ(TrackStatus::initializing, sim.status());
+
+    // Run pre-step, set the state to something bad, check that it throws
+    id = this->find_action("pre-step");
+    this->run(id, state);
+    EXPECT_EQ(TrackStatus::alive, sim.status());
+    sim.status(TrackStatus::initializing);
+    this->check_throw(id, state, "status was improperly reverted");
+    // ...and restore the state
+    sim.status(TrackStatus::alive);
+
+    // Run kernel to select a model
+    id = this->find_action("physics-discrete-select");
+    this->run(id, state);
+
+    // Set a bogus along-step state
+    sim.along_step_action(this->find_action("pre-step"));
+    this->check_throw(id, state, "along-step action cannot yet change");
+    sim.along_step_action(this->find_action("along-step-neutral"));
+
+    // Run klein-nishina, change the ID *back* to an earlier action
+    id = this->find_action("scat-klein-nishina");
+    this->run(id, state);
+    EXPECT_EQ("scat-klein-nishina", this->id_to_label(sim.post_step_action()));
+    sim.post_step_action(this->find_action("physics-discrete-select"));
+    this->check_throw(id, state, "new post-step action is out of order");
+}
+
+TEST_F(StatusCheckerTest, TEST_IF_CELER_DEVICE(device))
```

Comment:
I'm seeing a failure for this test:
```console
[ RUN      ] StatusCheckerTest.device
/home/alund/celeritas_project/celeritas/src/celeritas/phys/PhysicsParams.cc:337: warning: No processes are defined for particle 'electron'
status: Celeritas core setup complete
status: Celeritas core state initialization complete
/home/alund/celeritas_project/celeritas/src/celeritas/track/detail/StatusCheckExecutor.hh:119:
celeritas: internal assertion failed: new post-step action is out of order: '(prev_order == params.implicit_order || next_order == params.implicit_order || OrderedAction{prev_order, prev_post_step} < OrderedAction{next_order, next_post_step})' failed
/home/alund/celeritas_project/celeritas/test/celeritas/track/StatusChecker.test.cc:145: Failure
Value of: actual_message.find(match) != std::string::npos
  Actual: false
Expected: true
Actual message: 'unspecified launch failure'
/home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:129: debug: While freeing device memory: celeritas: CUDA error: unspecified launch failure
/home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:118: 'cudaFree(ptr)' failed
/home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:129: debug: While freeing device memory: celeritas: CUDA error: unspecified launch failure
/home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:118: 'cudaFree(ptr)' failed
/home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:133: debug: Suppressing further DeviceFreeDeleter warning messages
Writing diagnostic output because test failed
info: Wrote output to statuschecker-device-1.out.json
[  FAILED  ] StatusCheckerTest.device (44 ms)
```

Reply:
Maybe it's too much to require that CUDA produces an expected error message from an assertion failure: that probably gets into "undefined behavior" territory. Maybe you could replace:
```c++
    this->check_throw(this->find_action("scat-klein-nishina"),
                      state,
                      "device-side assert triggered");
```
with
```c++
    EXPECT_THROW(
        status_checker_->execute(this->find_action("scat-klein-nishina"), *this->core(), state),
        RuntimeError);
```
in #1307 ?